donkeycar/parts/pins.py

```diff
@@ -0,0 +1,913 @@
+
+    raise ValueError("Unknown pin provider {}".format(parts[0]))
+
+
+def pwm_pin_by_id(pin_id:str, frequency_hz:int=60) -> PwmPin:
+    """
+    Select a pwm output pin given a pin id.
+    """
+    parts = pin_id.split(".")
+    if parts[0] == PinProvider.PCA9685:
+        pin_provider = parts[0]
+        i2c_bus, i2c_address = parts[1].split(":")
+        i2c_bus = int(i2c_bus)
+        i2c_address = int(i2c_address, base=16)
+        pin_number = int(parts[2])
+        return pwm_pin(pin_provider, pin_number, i2c_bus=i2c_bus, i2c_address=i2c_address, frequency_hz=frequency_hz)
+
+    if parts[0] == PinProvider.RPI_GPIO:
+        pin_provider = parts[0]
+        pin_scheme = parts[1]
+        pin_number = int(parts[2])
+        return pwm_pin(pin_provider, pin_number, pin_scheme=pin_scheme, frequency_hz=frequency_hz)
+
+    if parts[0] == PinProvider.PIGPIO:
+        pin_provider = parts[0]
+        if PinScheme.BCM != parts[1]:
+            raise ValueError("Pin scheme must be BCM for PIGPIO")
+        pin_number = int(parts[2])
+        return pwm_pin(pin_provider, pin_number, pin_scheme=PinScheme.BCM, frequency_hz=frequency_hz)
+
+    raise ValueError("Unknown pin provider {}".format(parts[0]))
+
+
+def input_pin_by_id(pin_id:str, pull:int=PinPull.PULL_NONE) -> InputPin:
+    """
+    Select a ttl input pin given a pin id.
+    """
+    parts = pin_id.split(".")
+    if parts[0] == PinProvider.PCA9685:
+        raise RuntimeError("PinProvider.PCA9685 does not implement InputPin")
+
+    if parts[0] == PinProvider.RPI_GPIO:
+        pin_provider = parts[0]
+        pin_scheme = parts[1]
+        pin_number = int(parts[2])
+        return input_pin(pin_provider, pin_number, pin_scheme=pin_scheme, pull=pull)
+
+    if parts[0] == PinProvider.PIGPIO:
+        pin_provider = parts[0]
+        if PinScheme.BCM != parts[1]:
+            raise ValueError("Pin scheme must be BCM for PIGPIO")
+        pin_number = int(parts[2])
+        return input_pin(pin_provider, pin_number, pin_scheme=PinScheme.BCM, pull=pull)
+
+    raise ValueError("Unknown pin provider {}".format(parts[0]))
+
+
+def input_pin(pin_provider:str, pin_number:int, pin_scheme:str=PinScheme.BOARD, pull:int=PinPull.PULL_NONE) -> InputPin:
+    """
+    construct an InputPin using the given pin provider
+    """
+    if pin_provider == PinProvider.RPI_GPIO:
+        return InputPinGpio(pin_number, pin_scheme, pull)
+    if pin_provider == PinProvider.PCA9685:
+        raise RuntimeError("PinProvider.PCA9685 does not implement InputPin")
+    if pin_provider == PinProvider.PIGPIO:
+        if pin_scheme != PinScheme.BCM:
+            raise ValueError("Pin scheme must be PinScheme.BCM for PIGPIO")
+        return InputPinPigpio(pin_number, pull)
+    raise RuntimeError("UnknownPinProvider ({})".format(pin_provider))
+
+
+def output_pin(pin_provider:str, pin_number:int, pin_scheme:str=PinScheme.BOARD, i2c_bus:int=0, i2c_address:int=40, frequency_hz:int=60) -> OutputPin:
```

Comment:
Rather lengthy line... we should probably try to stick to pep-8.

Reply:
I'm doing my linting with 120 character line length.